### ISSUE TITLE
Relative URLs Broken in News after Upgrade

### ISSUE THREAD

**Issue by miqrogroove** [fisharebest/webtrees#3207]
I thought this was supposed to work automatically, but it's not working.  Paths like..

individual.php?pid=I1962&ged=tree1

... are not redirecting from old News articles, because they are now apparently relative to /tree/ intead of the site root.  When the browser hits /tree/individual.php?pid=I1962&ged=tree1 then the site just dumbly redirects me to /tree/tree1

I'm planning to manually adjust all of my links, but I thought you should be aware that this is a significant problem after upgrading.

**Comment by fisharebest** [comment]
We might be able to fix this by adding a `<base>` tag in the HTML header.

**Comment by miqrogroove** [comment]
Changing the base would just break all the new relative links for the advantage of the old ones.

I would suggest instead trying to make the /tree/ path routable/forwardable as a fallback for old news links.

And if that doesn't work, perhaps you could offer one of those "data fix" menus for re-basing any `href="individual.php..."` and similar things in the news.

**Comment by fisharebest** [comment]
> Changing the base would just break all the new relative links for the advantage of the old ones.

Looking at the documentation, it would also break any in-page links, such as `<a href="#something">`

**Comment by fisharebest** [comment]
> I would suggest instead trying to make the /tree/ path routable/forwardable as a fallback

We also need `/tree/<tree>/my-page/individual.php` for links on the user's "my page".

Adding lots of new URLs will have an effect on performance.  There was a recent request on the forum to match *all* the old webtrees/phpGedView URLs (charts, reports, etc.).

See https://www.webtrees.net/index.php/en/forum/help-for-2-0/34396-redirection-of-wt-1-urls-to-wt-2-syntax

This will need to be a module, so that it can be disabled by users who do not need it.

**Comment by miqrogroove** [comment]
> Adding lots of new URLs will have an effect on performance. There was a recent request on the forum to match _all_ the old webtrees/phpGedView URLs (charts, reports, etc.).

For what it's worth, these URLs only need to be tested as a fallback when the new namespace doesn't match anything.  So when the /tree/... route fails to find the right page, let it check for "old" URLs.  In that way it would only impact performance on the "old" links.

**Comment by fisharebest** [comment]
> these URLs only need to be tested as a fallback when the new namespace doesn't match anything

I tried coding this.

The problem is that the URL *does* match an existing rule.

If you are viewing `/tree/demo` and you follow a relative link to `individual.php`, you will arrive at `/tree/individual.php`, which is matched by the route `/tree/{tree}`.

**Comment by miqrogroove** [comment]
Is it trying to find a tree named individual.php?  And if so, couldn't any failure to find the tree trigger a check then for old URLs?

**Comment by fisharebest** [comment]
> Is it trying to find a tree named individual.php? 

Yes.  A tree name can be any valid filename.  `individual.php` is a valid tree name.

I agree this is unlikely, but it is valid.

> couldn't any failure to find the tree trigger a check then for old URLs?

By design, the code cannot distinguish between a private tree and a tree that does not exist.

(Yes - we can always make an extra database query to bypass the privacy checks, but I would prefer not to do this.)

**Comment by miqrogroove** [comment]
Even if you insist on allowing individual.php to be a "valid" tree name, the extra query only needs to occur in that specific situation.  The performance impact would be almost zero.
